connection_test.go

```diff
@@ -868,3 +868,30 @@ func TestTosPriority(t *testing.T) {
 		require.NoError(t, err, "Failed to write to outbound conn")
 	})
 }
+
+func TestPeerStatusChangeOnDisconnect(t *testing.T) {
+	testutils.WithTestServer(t, nil, func(ts *testutils.TestServer) {
+		server := ts.Server()
+		testutils.RegisterEcho(server, nil)
+		changes := make(chan int, 2)
+
+		copts := testutils.NewOpts().SetOnPeerStatusChanged(func(p *Peer) {
+			i, o := p.NumConnections()
+			changes <- i + o
+		})
+
+		// Induce the creation of a connection from client to server.
```

Comment:
can we ensure that we get a notification on the server for all clients by:
 - setting the `SetOnPeerStatusChanged` on the server
 - using a for loop with multiple clients, and ensuring that each new client triggers the event

```go
for i := 0; i < 5; i++ {
  client := ts.NEwClient(copts)
  testutils.AssertEcho(t, client, ts.HostPort(), ts.ServiceName())
  assert.Equal(t, i, <-changes, "connected")

  // reuse connection
  testutils.AssertEcho(t, client, ts.HostPort(), ts.ServiceName())
  assert.Len(t, 0, changes, "reusing connection should not trigger any events")
}
```
we should also verify that there are no queued events after we assert echo a second time.

we'll have to disable the delay as an option.